primitives/account/src/account/staking_contract/staker.rs

```diff
@@ -316,25 +338,38 @@ impl StakingContract {
     ) -> Result<(), AccountError> {
         // Get the staker.
         let mut staker = store.expect_staker(staker_address)?;
+        let total_balance = staker.balance + staker.inactive_balance;
 
         // Remove ourselves from the current delegation, if it exists.
-        if staker.delegation.is_some() && !staker.balance.is_zero() {
-            self.remove_staker_from_validator(store, &staker)
-                .expect("inconsistent contract state");
+        if let Some(validator_address) = &staker.delegation {
+            if receipt.reactivate_all_stake {
+                self.decrease_stake_from_validator(store, validator_address, staker.balance)
+            }
```

Comment:
Same here, but in reverse. When unregistering a staker from a validator, its balance must always be decreased, too. I'd remove the condition here and always call it. Cannot hurt.